src/user_input/mod.rs

```diff
@@ -0,0 +1,424 @@
+///
+/// # Examples
+///
+/// ```rust
+/// use std::hash::{Hash, Hasher};
+/// use bevy::prelude::*;
+/// use bevy::utils::FloatOrd;
+/// use serde::{Deserialize, Serialize};
+/// use leafwing_input_manager::prelude::*;
+/// use leafwing_input_manager::input_streams::InputStreams;
+/// use leafwing_input_manager::axislike::{DualAxisType, DualAxisData};
+/// use leafwing_input_manager::raw_inputs::RawInputs;
+/// use leafwing_input_manager::clashing_inputs::BasicInputs;
+///
+/// #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Reflect, Serialize, Deserialize)]
+/// pub struct MouseScrollAlwaysFiveOnYAxis;
+///
+/// // Add this attribute for ensuring proper serialization and deserialization.
+/// #[serde_typetag]
+/// impl UserInput for MouseScrollAlwaysFiveOnYAxis {
+///     fn kind(&self) -> InputControlKind {
+///         // Returns the kind of input this represents.
+///         //
+///         // In this case, it represents an axial input.
+///         InputControlKind::Axis
+///     }
+///
+///     fn pressed(&self, input_streams: &InputStreams) -> bool {
+///         // Checks if the input is currently active.
+///         //
+///         // Since this virtual mouse scroll always outputs a value,
+///         // it will always return `true`.
+///         true
+///     }
+///
+///     fn value(&self, input_streams: &InputStreams) -> f32 {
+///         // Gets the current value of the input as an `f32`.
+///         //
+///         // This input always represents a scroll of `5.0` on the Y-axis.
+///         5.0
+///     }
+///
+///     fn axis_pair(&self, input_streams: &InputStreams) -> Option<DualAxisData> {
+///         // Gets the values of this input along the X and Y axes (if applicable).
+///         //
+///         // This input only represents movement on the Y-axis,
+///         // so it returns `None`.
+///         None
+///     }
+///
+///     fn decompose(&self) -> BasicInputs {
+///         // Gets the most basic form of this input for clashing input detection.
+///         //
+///         // This input is a simple, atomic unit,
+///         // so it is returned as a `BasicInputs::Simple`.
+///         BasicInputs::Simple(Box::new(*self))
+///     }
+///
+///     fn raw_inputs(&self) -> RawInputs {
+///         // Defines the raw input events used for simulating this input.
+///         //
+///         // This input simulates a mouse scroll event on the Y-axis.
+///         RawInputs::from_mouse_scroll_axes([DualAxisType::Y])
+///     }
+/// }
+///
+/// // Remember to register your input - it will ensure everything works smoothly!
+/// let mut app = App::new();
+/// app.register_user_input::<MouseScrollAlwaysFiveOnYAxis>();
+/// ```
+pub trait UserInput:
+    Send + Sync + Debug + DynClone + DynEq + DynHash + Reflect + erased_serde::Serialize
+{
+    /// Defines the kind of behavior that the input should be.
+    fn kind(&self) -> InputControlKind;
```

Comment:
I think that the better design here is to have three separate traits `ButtonlikeUserInput` etc that all rely on a core `UserInput` trait that captures shared behavior. I think that's likely to be easier to do in a follow-up though.

Reply:
I had previously considered splitting it into different input kinds: Button, Axis, DualAxis, etc.; and there are corresponding data types for each. Then, modify the ActionState API, something like this:
```
if let ButtonState(state) = action_state.get(ButtonAction) {
   let pressed = state.pressed;
   let pressure = state.pressure; // Gamepad buttons are ranging from 0 to 1, and others are 0 or 1
}
if let AxisValue(value: f32) = action_state.get(AxisAction) {
}
```